File: test/velocity_constraint/velocity_constraint.py
from minieigen import *
from pyrscl import *
from math import *
import sys

def isClose(v1, v2, eps = 1e-3):
	return abs(v1-v2) < eps

def setZero(vec):
	for i in range(0,6):
		vec[i] = 0

def getTransVel(vec):
	t = Vector3.Zero
	for i in range(0,3):
		t[i] = vec[i]
	return t

damping_matrix = NewMatrix6dPtr(Matrix6.Identity)
maximum_velocity = NewDoublePtr(0.5)

safety_controller = NewSafetyController(damping_matrix)
safety_controller.setVerbose(True)
tcp_velocity = safety_controller.getTCPVelocity()
total_velocity = safety_controller.getTotalVelocity()

velocity_constraint = NewVelocityConstraint(maximum_velocity)

constant_vel = NewVector6dPtr()
constant_velocity_generator = NewVelocityProxy(constant_vel)

safety_controller.addConstraint("velocity constraint", velocity_constraint)
safety_controller.addVelocityGenerator("vel proxy", constant_velocity_generator)

# Step #1 : no velocity
safety_controller.updateTCPVelocity()

assert_msg("Step #1", tcp_velocity.isApprox(Vector6.Zero))

# Step #2 : velocity 1 axis < max
constant_vel[0] = 0.2
safety_controller.updateTCPVelocity()

assert_msg("Step #2", tcp_velocity.isApprox(total_velocity))

# Step #3 : velocity 1 axis > max
constant_vel[0] = 0.6
safety_controller.updateTCPVelocity()

assert_msg("Step #3", isClose(getTransVel(tcp_velocity).norm(), maximum_velocity.get()))

# Step #4 : velocity 3 axes < max
constant_vel[0] = 0.2
constant_vel[1] = 0.1
constant_vel[2] = 0.3
safety_controller.updateTCPVelocity()

assert_msg("Step #4", tcp_velocity.isApprox(total_velocity))

# Step #5 : velocity 3 axes > max
constant_vel[0] = 0.5
constant_vel[1] = 0.4
constant_vel[2] = 0.6
safety_controller.updateTCPVelocity()

assert_msg("Step #5", isClose(getTransVel(tcp_velocity).norm(), maximum_velocity.get()))

# Step #6 : rotational velocity only
setZero(constant_vel)
constant_vel[4] = 0.5
constant_vel[4] = 0.4
constant_vel[5] = 0.6
safety_controller.updateTCPVelocity()

assert_msg("Step #6", tcp_velocity.isApprox(total_velocity))

sys.exit(0)

File: test/velocity_constraint/velocity_constraint.cpp
#undef NDEBUG

#include <OpenPHRI/OpenPHRI.h>

using namespace phri;

bool isClose(double v1, double v2, double eps = 1e-3) {
    return std::abs(v1 - v2) < eps;
}

int main(int argc, char const* argv[]) {

    auto robot = make_shared<Robot>("rob", // Robot's name
                                    7);    // Robot's joint count

    auto safety_controller = SafetyController(robot);
    safety_controller.setVerbose(true);

    auto maximum_velocity = make_shared<double>(0.5);
    auto velocity_constraint =
        make_shared<VelocityConstraint>(maximum_velocity);

    auto constant_vel = make_shared<Twist>();
    auto constant_velocity_generator = make_shared<VelocityProxy>(constant_vel);

    safety_controller.add("velocity constraint", velocity_constraint);
    safety_controller.add("vel proxy", constant_velocity_generator);

    const Vector6d& cp_velocity = *robot->controlPointVelocity();
    const Vector6d& cp_total_velocity = *robot->controlPointTotalVelocity();

    // Step #1 : no velocity
    safety_controller.compute();

    assert_msg("Step #1", cp_velocity.isZero());

    // Step #2 : velocity 1 axis < max
    constant_vel->translation().x() = 0.2;
    safety_controller.compute();

    assert_msg("Step #2", cp_velocity.isApprox(cp_total_velocity));

    // Step #3 : velocity 1 axis > max
    constant_vel->translation().x() = 0.6;
    safety_controller.compute();

    assert_msg("Step #3",
               isClose(robot->controlPointVelocity()->translation().norm(),
                       *maximum_velocity));

    // Step #4 : velocity 3 axes < max
    constant_vel->translation().x() = 0.2;
    constant_vel->translation().y() = 0.1;
    constant_vel->translation().z() = 0.3;
    safety_controller.compute();

    assert_msg("Step #4", cp_velocity.isApprox(cp_total_velocity));

    // Step #5 : velocity 3 axes > max
    constant_vel->translation().x() = 0.5;
    constant_vel->translation().y() = 0.4;
    constant_vel->translation().z() = 0.6;
    safety_controller.compute();

    assert_msg("Step #5",
               isClose(robot->controlPointVelocity()->translation().norm(),
                       *maximum_velocity));

    // Step #6 : rotational velocity only
    static_cast<Vector6d&>(*constant_vel).setZero();
    constant_vel->rotation().x() = 0.5;
    constant_vel->rotation().y() = 0.4;
    constant_vel->rotation().z() = 0.6;
    safety_controller.compute();

    assert_msg("Step #6", cp_velocity.isApprox(cp_total_velocity));

    return 0;
}
